src/stan/services/optimize/lbfgs.hpp

```diff
@@ -0,0 +1,206 @@
+       * @param random_seed random seed for the pseudo random number generator
+       * @param chain chain id to advance the pseudo random number generator
+       * @param init_radius radius to initialize
+       * @param[in] history_size amount of history to keep for L-BFGS
+       * @param[in] init_alpha line search step size for first iteration
+       * @param[in] tol_obj convergence tolerance on absolute changes in
+       *   objective function value
+       * @param[in] tol_rel_obj convergence tolerance on relative changes
+       *   in objective function value
+       * @param[in] tol_grad convergence tolerance on the norm of the gradient
+       * @param[in] tol_rel_grad convergence tolerance on the relative norm of
+       *   the gradient
+       * @param[in] tol_param convergence tolerance on changes in parameter value
+       * @param[in] num_iterations maximum number of iterations
+       * @param[in] save_iterations indicates whether all the interations should
+       *   be saved to the parameter_writer
+       * @param[in] refresh how often to write output to message_writer
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       * @return stan::services::error_codes::OK if successful
+       */
+      template <class Model>
+      int lbfgs(Model& model,
+                stan::io::var_context& init,
+                unsigned int random_seed,
+                unsigned int chain,
+                double init_radius,
+                int history_size,
+                double init_alpha,
+                double tol_obj,
+                double tol_rel_obj,
+                double tol_grad,
+                double tol_rel_grad,
+                double tol_param,
+                int num_iterations,
+                bool save_iterations,
+                int refresh,
+                callbacks::interrupt& interrupt,
+                callbacks::writer& message_writer,
+                callbacks::writer& init_writer,
+                callbacks::writer& parameter_writer) {
+        boost::ecuyer1988 rng = stan::services::util::rng(random_seed, chain);
+
+        std::vector<int> disc_vector;
+        std::vector<double> cont_vector
+          = stan::services::util::initialize(model, init, rng, init_radius,
+                                             false,
+                                             message_writer, init_writer);
+
+        std::stringstream lbfgs_ss;
+        typedef stan::optimization::BFGSLineSearch
+          <Model, stan::optimization::LBFGSUpdate<> > Optimizer;
+        Optimizer lbfgs(model, cont_vector, disc_vector, &lbfgs_ss);
+        lbfgs.get_qnupdate().set_history_size(history_size);
+        lbfgs._ls_opts.alpha0 = init_alpha;
```

Comment:
From CmdStan, RStan, PyStan. I'm pretty sure that's what we decided. Personally, I would have wanted to have (redundant) validation inside the service functions, but I think at some point I was convinced that it was better practice to validate arguments outside.